subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ForExternalUseTest.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.architecture.test;
+
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchRule;
+import org.gradle.internal.accesscontrol.AllowUsingApiForExternalUse;
+import org.gradle.internal.accesscontrol.ForExternalUse;
+
+import static com.tngtech.archunit.base.DescribedPredicate.or;
+import static com.tngtech.archunit.core.domain.properties.CanBeAnnotated.Predicates.annotatedWith;
+import static com.tngtech.archunit.lang.conditions.ArchConditions.and;
+import static com.tngtech.archunit.lang.conditions.ArchConditions.beDeclaredInClassesThat;
+import static com.tngtech.archunit.lang.conditions.ArchConditions.beProtected;
+import static com.tngtech.archunit.lang.conditions.ArchConditions.bePublic;
+import static com.tngtech.archunit.lang.conditions.ArchConditions.or;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.codeUnits;
+import static org.gradle.architecture.test.ArchUnitFixture.gradlePublicApi;
+
+@AnalyzeClasses(packages = "org.gradle")
+public class ForExternalUseTest {
+
+    private static final Class<ForExternalUse> FOR_EXTERNAL_USE_ANNOTATION = ForExternalUse.class;
+
+    @ArchTest
+    public static final ArchRule members_for_external_use_are_public_api =
+        codeUnits().that().areAnnotatedWith(FOR_EXTERNAL_USE_ANNOTATION)
+            .should(
+                and(
+                    beDeclaredInClassesThat(gradlePublicApi()),
+                    or(bePublic(), beProtected())
+                )
+            );
+
+    @ArchTest
+    public static final ArchRule members_for_external_use_are_not_used_internally =
+        codeUnits().that().areAnnotatedWith(FOR_EXTERNAL_USE_ANNOTATION)
+            .should().onlyBeCalled().byCodeUnitsThat(
+                or(
+                    annotatedWith(FOR_EXTERNAL_USE_ANNOTATION),
+                    annotatedWith(AllowUsingApiForExternalUse.class)
+                )
+            );
```

Comment:
I prefer using the fluent API, so something like this:
```suggestion
    @ArchTest
    public static final ArchRule members_for_external_use_are_public_api =
        codeUnits().that().areAnnotatedWith(FOR_EXTERNAL_USE_ANNOTATION)
            .should(beDeclaredInClassesThat(are(gradlePublicApi())))
            .andShould(bePublic().or(beProtected()));

    @ArchTest
    public static final ArchRule members_for_external_use_are_not_used_internally =
        codeUnits().that().areAnnotatedWith(FOR_EXTERNAL_USE_ANNOTATION)
            .should().onlyBeCalled().byCodeUnitsThat(are(
                    annotatedWith(FOR_EXTERNAL_USE_ANNOTATION).or(annotatedWith(AllowUsingApiForExternalUse.class))
                )
            );
```

The phrase will also be printed when the test fails, so using things like `ArchPredicates.are` is a good idea.

Reply:
Thanks you! Accepted it.